pkg/services/ngalert/state/historian/loki_test.go

```diff
@@ -250,6 +259,69 @@ func TestMerge(t *testing.T) {
 	}
 }
 
+func TestRecordStates(t *testing.T) {
+	t.Run("writes state transitions to loki", func(t *testing.T) {
+		req := NewFakeRequester()
+		loki := createTestLokiBackend(req, metrics.NewHistorianMetrics(prometheus.NewRegistry()))
+		rule := createTestRule()
+		states := singleFromNormal(&state.State{
+			State:  eval.Alerting,
+			Labels: data.Labels{"a": "b"},
+		})
+
+		err := <-loki.RecordStatesAsync(context.Background(), rule, states)
+
+		require.NoError(t, err)
+		require.Contains(t, "/loki/api/v1/push", req.lastRequest.URL.Path)
+	})
+
+	t.Run("emits expected write metrics", func(t *testing.T) {
+		reg := prometheus.NewRegistry()
+		met := metrics.NewHistorianMetrics(reg)
+		loki := createTestLokiBackend(NewFakeRequester(), met)
+		errLoki := createTestLokiBackend(NewFakeRequester().WithResponse(badResponse()), met) //nolint:bodyclose
+		rule := createTestRule()
+		states := singleFromNormal(&state.State{
+			State:  eval.Alerting,
+			Labels: data.Labels{"a": "b"},
+		})
+
+		<-loki.RecordStatesAsync(context.Background(), rule, states)
+		<-errLoki.RecordStatesAsync(context.Background(), rule, states)
+
+		exp := bytes.NewBufferString(`
+# HELP grafana_alerting_state_history_transitions_failed_total The total number of state transitions that failed to be written - they are not retried.
+# TYPE grafana_alerting_state_history_transitions_failed_total counter
+grafana_alerting_state_history_transitions_failed_total{org="1"} 1
+# HELP grafana_alerting_state_history_transitions_total The total number of state transitions processed by the state historian.
+# TYPE grafana_alerting_state_history_transitions_total counter
+grafana_alerting_state_history_transitions_total{org="1"} 2
+# HELP grafana_alerting_state_history_writes_failed_total The total number of failed writes of state history batches.
+# TYPE grafana_alerting_state_history_writes_failed_total counter
+grafana_alerting_state_history_writes_failed_total{org="1"} 1
+# HELP grafana_alerting_state_history_writes_total The total number of state history batches that were attempted to be written.
+# TYPE grafana_alerting_state_history_writes_total counter
+grafana_alerting_state_history_writes_total{org="1"} 2
+`)
+		err := testutil.GatherAndCompare(reg, exp,
+			"grafana_alerting_state_history_transitions_total",
+			"grafana_alerting_state_history_transitions_failed_total",
+			"grafana_alerting_state_history_writes_total",
+			"grafana_alerting_state_history_writes_failed_total",
```

Comment:
```suggestion
			"grafana_alerting_state_history_transitions_total",
			"grafana_alerting_state_history_transitions_failed_total",
			"grafana_alerting_state_history_writes_total",
			"grafana_alerting_state_history_writes_failed_total",
			"grafana_alerting_state_history_writes_bytes_total"
```

Reply:
Did we missed writes here?

Reply:
This was done intentionally - since the log lines contain timestamps, the expected number of bytes is variable and not deterministic. I chose to leave it out for now to avoid test intermittency.

We can solve this by introducing something like `benbjohnson/clock` to the Loki layer, so we can have deterministic times in tests, and therefore deterministic number of bytes. I'd really prefer to do this in a separate PR.

Reply:
Not a problem at all.